src/sdsc/spellcheck.py

```diff
@@ -0,0 +1,232 @@
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+#
+
+import re
+import enchant
+import os.path
+from lxml import etree
+
+# remove this line later, we only need it for sentencesegmenterREMOVEME
+from .const import PARENTHESES, SENTENCEENDS, LASTSENTENCEENDS, EMPTYSUBPATTERN
+
+from .const import SPELLFILTER
+from .cli import printcolor
+from .generic import (
+        linenumber,
+        re_compile,
+                     )
+from .textutil import (
+        findtagreplacement,
+        removepunctuation,
+        sanitizepunctuation,
+        tokenizer,
+        xmlescape,
+                      )
+
+
+def spellcheck(context, maindict, customdict, content, contentpretty,
+               contextid, basefile, messagetype):
+    """ Check a paragraph using using enchant's default spell checker. If there
+    is a misspelling, issue a message.
+
+    :param ??? context: information about the context node
+    :param str maindict: system-wide main dictionary to use
+    :param str customdict: extra word list to use (optional)
+    :param str content: content, as formatted for the terminology check itself
+    :param str contentpretty: content, as formatted for display in a message
+    :param str contextid: next element with id attribute around the content
+    :param str basefile: file in which content appears
+    :param str messagetype: print a 'warning', 'info', or 'error' message?
+    """
+
+    # FIXME: Much of this is copypasta from termcheck(), though termcheck()
+    # is much more complicated and a lot of conditions are removed in this
+    # version. Refactor?
+
+    if not content:
+        return []
+
+    # I get this as a list with one lxml.etree._ElementUnicodeResult.
+    # I need a single string.
+    # For whatever reason, this made spellcheckmessage() crash
+    # happily and semi-randomly.
+    content = sanitizepunctuation(str(content[0]), quotes=False, apostrophes=True)
+
+    basefile = basefile[0] if basefile else None
+    contextid = contextid[0] if contextid else None
+
+    # sanitize this...
+    if messagetype not in ('warning', 'info'):
+        messagetype = 'error'
+
+    # This if/else block should not be necessary (if there is content,
+    # there should always also be pretty content, but that depends on the
+    # XSLT used for checking). It hopefully won't hurt either.
+    contentpretty = str(contentpretty[0]) if contentpretty else content
+
+    # FIXME: profile this for speed -- we might have to make this global(ler).
+    # Also, the approach o
+    # need to convert byte string to string
+    spelldict = 0
+    if maindict and customdict:
+        customdict = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'xsl-checks', str(customdict))
+        spelldict = enchant.DictWithPWL(str(maindict), customdict)
+    elif maindict:
+        spelldict = enchant.Dict(str(maindict))
+    else:
+        printcolor
```

Comment:
I'm not sure if I understand this line. For me it doesn't make sense. Is it a function call? In that case, it's a typo and brackets are missing.

Reply:
yeah, i forgot something there...